### PR TITLE
Make it so that async-fn-in-trait is compatible with a concrete future in implementation

### PR DESCRIPTION
**Pull request by compiler-errors**
There's no technical reason why an AFIT like `async fn foo()` cannot be satisfied with an implementation signature like `fn foo() -> Pin<Box<dyn Future<Output = ()> + 'static>>`.

We rejected this previously because we were uncertain about how AFITs worked with refinement, but I don't believe this needs to be a restriction any longer.

r? oli-obk

### CODE REVIEW

**Comment by compiler-errors** [comment]
I'm gonna lang nominate this.

@rustbot label: I-lang-nominated

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs [hunk 1]
```diff
@@ -663,8 +663,14 @@ pub(super) fn collect_return_position_impl_trait_in_trait_tys<'tcx>(
     // RPITs.
     let errors = ocx.select_all_or_error();
     if !errors.is_empty() {
-        let reported = infcx.err_ctxt().report_fulfillment_errors(errors);
-        return Err(reported);
+        if let Some(guar) =
+            try_report_async_mismatch(tcx, infcx, &errors, trait_m, impl_m, impl_sig)
```
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:667]: change `try_report_async_mismatch` to return `Result<(), ErrorGuaranteed>` and use `?` here.

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:667]: 🧠

**Comment by tmandry** [comment]
We discussed this in the lang team meeting; everyone present was in favor, but agreed that this was a form of refinement and should be treated like any other kind of refinement.

@rfcbot merge

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

As **tmandry** noted, we discussed this in T-lang triage and were in favor of doing this.  It's now entering FCP, so let's unnominate.

**Comment by matthiaskrgr** [comment]
Maybe failed in  https://github.com/rust-lang/rust/pull/120653#issuecomment-1925977937 ? (not sure)

**Comment by compiler-errors** [comment]
oops forgot to re-approve

@bors r=oli-obk

File: compiler/rustc_hir_analysis/src/check/compare_impl_item.rs [hunk 2]
```diff
@@ -695,8 +664,13 @@ pub(super) fn collect_return_position_impl_trait_in_trait_tys<'tcx>(
     // RPITs.
     let errors = ocx.select_all_or_error();
     if !errors.is_empty() {
-        let reported = infcx.err_ctxt().report_fulfillment_errors(errors);
-        return Err(reported);
+        if let Err(guar) = try_report_async_mismatch(tcx, infcx, &errors, trait_m, impl_m, impl_sig)
```
Comment (Cheban1996) [inline compiler/rustc_hir_analysis/src/check/compare_impl_item.rs:667]: I think this line better ```try_report_async_mismatch(tcx, infcx, &errors, trait_m, impl_m, impl_sig)?;``` than now

**Pull request merged**